### ISSUE TITLE
[Bug Report] 1.41.3 Failure after receiving twin responses from module still occurs

### ISSUE THREAD

**Issue by sevywzgarda** [Azure/azure-iot-sdk-csharp#3044]
# Context

- **OS :** Ubuntu Server 18.04.6 LTS / Amd64
- **Application's .NET Target Framework :** dotnet 6.0
- **Device :** VM (VMware)
- **SDK version used :** Microsoft.Azure.Devices.Client" Version=1.41.3

## Description of the issue
After coming across this thread https://github.com/Azure/azure-iot-sdk-csharp/issues/2449 we updated the SDK with the newer 1.41.3 library and reworked all our modules. However, after deploying the modules to our environment, we noticed that the problem remains and we still get the `return MessageResponse.Completed` at the end of the M2M call. 

The problem occurs sporadically when the modules have a little data load (receiving/processing/sending), and more often when they are heavily loaded.

## Console log of the issue
 **Logs from `$edgeHub` [INF]:**
```
[WRN] - Received unknown feedback message from iot-device-01/Module with lock token 9dfad21a-eb6a-447e-9102-e5c74bb0c061 and status Complete. Abandoning message.
```
 **Logs from `$edgeHub` [DBG]:** 
```
[DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.ModuleEndpoint] - Sending 1 message(s) to module iot-device-01/Module.
[DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Sent message with correlation ID 9dfad21a-eb6a-447e-9102-e5c74bb0c061 to iot-device-01/Module
[DBG] [Microsoft.Azure.Devices.Edge.Hub.Amqp.ClientConnectionHandler] - Sending telemetry message to iot-device-01/Module
[DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Received feedback Complete for message 9dfad21a-eb6a-447e-9102-e5c74bb0c061 from device/module iot-device-01/Module
[DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Received feedback Complete for message 9dfad21a-eb6a-447e-9102-e5c74bb0c061 from device/module iot-device-01/Module
[WRN] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Received unknown feedback message from iot-device-01/Module with lock token 9dfad21a-eb6a-447e-9102-e5c74bb0c061 and status Complete. Abandoning message.
[DBG] [Microsoft.Azure.Devices.Edge.Hub.Amqp.LinkHandlers.SendingLinkHandler] - Sent message with id 9dfad21a-eb6a-447e-9102-e5c74bb0c061 to device iot-device-01/Module
```

FYI: @graysddoman & @ashC05

**Comment by tmahmood-microsoft** [comment]
Hi @sevywzgarda, thank you for reaching out. To further look into the issue, can you please provide a complete log file and if possible, a simplified code snippet you are using containing data load for the message.
Also, I see you are using AMQP protocol, are you also able to reproduce it with MQTT protocol?

**Comment by sevywzgarda** [comment]
Hi @tmahmood-microsoft! 

Thank you for looking into my case and sorry for the late reply, we just got back from holiday and were able to check what you suggested. It turns out that the M2M communication runs without abandoned messages when the MQTT protocol is enabled. 

It seems that the AMQP transport needs a similar fix as in MQTT.



**Logs:**

Here is attached [$edgeHub_Logs](https://github.com/Azure/azure-iot-sdk-csharp/files/10360447/edgeHub_Logs.txt) with "Abandoning message" at: 
- 13:03:51.759
- 13:04:15.769

**Code snippet:**

Here is the simplified code:

```
await moduleClient.SetInputMessageHandlerAsync("input_A", CustomMethod, moduleClient); 

public static async Task<MessageResponse> CustomMethod(Message message, object userContext)        
        { 
        // custom logic
        
        return MessageResponse.Completed; 
        }
```

**Comment by bqstony** [comment]
I god the same behavior and also use AMQP.

**Comment by tmahmood-microsoft** [comment]
Hi @sevywzgarda so far, I have not been able to reproduce the same warning message in my logs. 
Looking at your logs, I see the following logs regarding the same 'iot-device-01/Module' module which has 'Abandoning message' log.

2023-01-06 13:04:36.925 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.DeviceScopeAuthenticator] - Client iot-device-01/Module in device scope reauthenticated locally.
2023-01-06 13:04:36.925 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Authenticator] - Client iot-device-01/Module re-authentication successful
2023-01-06 13:04:36.925 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionReauthenticator] - Reauthenticated client iot-device-01/Module successfully

After these reauthentication logs, the abandoned message logs go away. 
Can you provide more details regarding the edge device(s) and the 3 modules listed in logs so far (Module, Module_pb, Module_hb)

Also, can you please check if MessageResponse.Completed is being sent twice for the specific module? The log would make sense if MessageResponse.Completed is being received twice for a message sent only once as the response is acknowledged twice for same message but for the second time, there is no message awaiting response and thus the warning message produced.

**Comment by graysddoman** [comment]
Hi @tmahmood-microsoft, 

For some context based on what @sevywzgarda has already shared. We have our modules setup with the routes as follows:

<img width="370" alt="image" src="https://user-images.githubusercontent.com/84330452/212319106-3825ee0f-7c26-4ab6-8302-ca6aa623cedd.png">

The modules which are highlighted are the ones that we have seen the warning message from. These are all internally developed modules and, from reviewing the code, follow the same pattern as provided in our code sample. So as far as I can tell, the MessageResponse.Completed is only sent once, after the message has been processed. 

We have tested with updated versions of the modules using MQTT as the protocol and it would appear the warnings disappear. So it seems that the issue manifests when we are using the AMQP protocol approach. Though we have observed that some of the modules do not exhibit this with certain combinations of the runtime and modules. I'll defer to @sevywzgarda for further details as he is closer to the actual deployment than I am.
